docs/proposals/20200220-cluster-resource-set.md

```diff
@@ -0,0 +1,210 @@
+
+### Implementation Details/Notes/Constraints
+
+#### Data model changes
+
+None. We are planning to implement this feature without modifying any of the existing structure to minimize the footprint of ClusterResourceSet Controller. This enhancement will follow Kubernetes’s feature-gate structure and will be under the experimental package with its APIs, and enabled/disabled with a feature gate. 
+
+#### ClusterResourceSet Object Definition
+
+This is the CRD that is used to have a set of components that will be applied to clusters that match the label selector in it.
+
+The resources field is a list of secrets/configmaps in the same namespace. The clusterSelector field is a Kubernetes [label selector](https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/#resources-that-support-set-based-requirements) that matches against labels on clusters (in any namespace).
+
+*Sample ClusterResourceSet YAML*
+
+```yaml
+---
+apiVersion: cluster.x-k8s.io/v1alpha3
+kind: ClusterResourceSet
+metadata:
+ name: postcreate-conf
+ namespace: default
+spec:
+ mode: "ApplyOnce"
+ clusterSelector:
+   matchLabels:
+     postcreatelabelcni: calico
+ resources:
+   - name: calico-addon
+     kind: Secret
+   - name: network-policy-addon
+     kind: ConfigMap
+status:
+ LastUpdated:  "2020-05-05T08:24:17Z"
+ StatusMessages: {
+                    {Message: "Added calico-addon resource", Type: "Update"},
+                    {Message: "Failed applying to x cluster", Type: "Error"},
+                }
+```
+
+Initially, the only supported mode will be `ApplyOnce` and it will be the default mode if no mode is provided. In the future, we may consider adding a `Sync` mode that reapplies the resources on resource hash change and/or periodically.
+If ClusterResourceSet resources will be managed by an operator after they are applied by ClusterResourceSet controller, "ApplyOnce" mode must be used so that reconciliation on those resources can be delegated to the operator.
+
+
+Each item in the resources specifies a kind (must be either ConfigMap or Secret) and a name. Each referenced ConfigMap/Secret  contains yaml/json content as value. The key to that content is “value”.
+
+*Sample Secret Format*
+```yaml
+apiVersion: v1
+kind: Secret
+metadata:
+  name:calico-addon
+type: Opaque
+stringData:
+  value: |-
+    kind: ConfigMap
+    apiVersion: v1
+    metadata:
+     name: calico-conf
+```
+
+The resources in ClusterResourceSet will be applied to matching clusters. 
+A configmap will be created in the management cluster to keep track of which resources are applied by ClusterResourceSet resources. There will be one configmap per workload cluster.
+
+Example:
+```yaml
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: capi-crs-my-cluster
+  namespace: capi-crs-my-cluster-namespace
+data:
+  <ClusterResourceSet-name1>:   
+    <secret-name1>:
+      hash: <>
+      status: success
+      error: ""
+      lastAppliedTime: "2020-04-05T08:24:17Z"
```

Comment:
Would the values of this ConfigMap be free-form json/yaml strings? If so, are we sure that we want to go down this path? When I originally suggested a configmap, I was looking for something very simple like `<id>`-> `<hash>`. Its existence would give us enough information to determine if something was already installed or not.


What do you all think?

Reply:
In kubeadm we went through this path (due to the lack of CRDs at that time), and we faced different shortcomings, like lack of validation, defaulting, conversion etc. 
Also, you don't have any control on the lifecycle of the config map itself

Reply:
In kubeadm we went through this path (due to the lack of CRDs at that time), and we faced different shortcomings, like lack of validation, defaulting, conversion etc.
Also, you don't have any control on the lifecycle of the config map itself

Reply:
I think we may need to keep track of more than simple key-value pairs in the future (even for the first version, we will need hash, status, error message as there is no better place to have these info.). So, to make things extensible for future improvements, a CRD would be a better solution than using configmaps. I will go with that approach, if we all agree.

Reply:
I'm +1 to a CRD. It also solves the naming question in https://github.com/kubernetes-sigs/cluster-api/pull/3050/files#r425564505 - we should have the names match.

Reply:
+1 to crd as well, but it should likely have a resource per ClusterResourceSet to avoid having multiple reconciliation threads attempting to modify the same resource in parallel.

Reply:
@ncdc do we need to potentially worry about name collisions on the "status" resource if we match names, particularly in the case that a ClusterResourceSet is deleted and recreated? I'm wondering if it would make sense to leverage GenerateName with a predictable prefix and handle the mapping of the "status" resource to the ClusterResourceSet with a field on the "status" resource and a label selector instead?

Reply:
Added a new CRD to keep track of which resources are applied to a cluster.